src/Temporalio/Bridge/CustomSlotSupplier.cs

```diff
@@ -0,0 +1,246 @@
+                try_reserve = FunctionPointer<Interop.CustomTryReserveSlotCallback>(TryReserve),
+                mark_used = FunctionPointer<Interop.CustomMarkSlotUsedCallback>(MarkUsed),
+                release = FunctionPointer<Interop.CustomReleaseSlotCallback>(Release),
+                free = FunctionPointer<Interop.CustomSlotImplFreeCallback>(Free),
+            };
+
+            PinCallbackHolder(interopCallbacks);
+        }
+
+        private static Temporalio.Worker.Tuning.SlotInfo SlotInfoFromBridge(Interop.SlotInfo slotInfo)
+        {
+            return slotInfo.tag switch
+            {
+                Interop.SlotInfo_Tag.WorkflowSlotInfo =>
+                    new Temporalio.Worker.Tuning.SlotInfo.WorkflowSlotInfo(
+                        ByteArrayRef.ToUtf8(slotInfo.workflow_slot_info.workflow_type), slotInfo.workflow_slot_info.is_sticky != 0),
+                Interop.SlotInfo_Tag.ActivitySlotInfo =>
+                    new Temporalio.Worker.Tuning.SlotInfo.ActivitySlotInfo(
+                        ByteArrayRef.ToUtf8(slotInfo.activity_slot_info.activity_type)),
+                Interop.SlotInfo_Tag.LocalActivitySlotInfo =>
+                    new Temporalio.Worker.Tuning.SlotInfo.LocalActivitySlotInfo(
+                        ByteArrayRef.ToUtf8(slotInfo.local_activity_slot_info.activity_type)),
+                _ => throw new System.ArgumentOutOfRangeException(nameof(slotInfo)),
+            };
+        }
+
+        private unsafe void Reserve(Interop.SlotReserveCtx* ctx, void* sender)
+        {
+            SafeReserve(new IntPtr(ctx), new IntPtr(sender));
+        }
+
+        // Note that this is always called by Rust, either because the call is cancelled or because
+        // it completed. Therefore the GCHandle is always freed.
+        private unsafe void CancelReserve(void* tokenSrc)
+        {
+            var handle = GCHandle.FromIntPtr(new IntPtr(tokenSrc));
+            var cancelTokenSrc = (CancellationTokenSource)handle.Target!;
+            cancelTokenSrc.Cancel();
+            handle.Free();
+        }
+
+        private void SafeReserve(IntPtr ctx, IntPtr sender)
+        {
+            _ = Task.Run(async () =>
+            {
+                using (var cancelTokenSrc = new System.Threading.CancellationTokenSource())
+                {
+                    unsafe
+                    {
+                        var srcHandle = GCHandle.Alloc(cancelTokenSrc);
+                        Interop.Methods.set_reserve_cancel_target(
+                            (Interop.SlotReserveCtx*)ctx.ToPointer(),
+                            GCHandle.ToIntPtr(srcHandle).ToPointer());
+                    }
+                    while (true)
+                    {
+                        try
+                        {
+                            ConfiguredTaskAwaitable<Temporalio.Worker.Tuning.SlotPermit> reserveTask;
```

Comment:
Never seen this type used. Docs say it's an internal compiler thing. I think it should be `Task<`.

Reply:
It won't compile as just `Task`

Reply:
Yeah, assuming it's because of the `unsafe` block, I'd say something like this should work:

```csharp
SlotReserveContext ctx;
unsafe { ctx = ReserveCtxFromBridge((Interop.SlotReserveCtx*)ctx.ToPointer()); }
var permit = await userSupplier.ReserveSlotAsync(ctx, cancelTokenSrc).ConfigureAwait(false);
```